### PR TITLE
introduce experimental compression codecs

### LINKED ISSUE #16775: Integrate and test experimental compression libraries.
See https://github.com/ClickHouse/ClickHouse/issues/15065#issuecomment-723397565
These codecs should be available under the experimental flag in Settings.

https://github.com/inikep/lizard
https://github.com/ConorStokes/LZSSE
https://github.com/centaurean/density
https://github.com/IlyaGrebnov/libbsc
https://github.com/fusiyuan2010/CSC
https://github.com/richgel999/lzham_codec

The task is for @fibersel

### PR DESCRIPTION
**Pull request by fibersel**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
(remove from changelog) Integrate and test experimental compression libraries. Will be available under the flag `allow_experimental_codecs`. This closes [#16775](https://github.com/ClickHouse/ClickHouse/issues/16775)

### CODE REVIEW

**Comment by fibersel** [comment]
@alexey-milovidov , Where can I find unit-tests for compression codecs?File `src/Compression/tests/gtest_compressionCodec.cpp` is not referenced in any target.

**Comment by alexey-milovidov** [comment]
@fibersel unit tests are built into single binary `unit_tests_dbms`.

**Review by alexey-milovidov — COMMENTED** [review]
Experimental codecs should also be protected by a setting:
`allow_experimental_codecs`, so users will not start to use it without being aware of the status of this feature.

**Comment by fibersel** [comment]
> Experimental codecs should also be protected by a setting:
> `allow_experimental_codecs`, so users will not start to use it without being aware of the status of this feature.

Well, I will account this condition before merging.
First step - implement and benchmark

**Comment by fibersel** [comment]
@alexey-milovidov can you suggest any example, that I can use to implement testing framework?
I saw `utils/compressor/decompress_perf.cpp`, but it seems to be too complicated for my case.

**Comment by alexey-milovidov** [comment]
@fibersel You can use `clickhouse-compressor` tool.

**Comment by alexey-milovidov** [comment]
How to use it:
1. When you will integrate your codecs to `CompressionCodecFactory`, they will become available under symbolic names in clickhouse-compressor.
2. You can use clickhouse-compressor to test compression speed and ratio. Then the same for decompression.

**Comment by fibersel** [comment]
@alexey-milovidov , how can I use it and measure time and compression ration?
I try `(base) [centos@clickhouse build]$ ./programs/clickhouse-compressor --stat  < URL.dat`
But get `Code: 39, e.displayText() = DB::Exception: Too large size_compressed. Most likely corrupted data., Stack trace (when copying this message, always include the lines below):

0. /home/centos/ClickHouse/contrib/poco/Foundation/src/Exception.cpp:27: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xcc0810c in /home/centos/build/programs/clickhouse
1. /home/centos/ClickHouse/src/Common/Exception.cpp:39: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x4309db1 in /home/centos/build/programs/clickhouse
2. /home/centos/ClickHouse/contrib/libcxx/include/string:2134: mainEntryClickHouseCompressor(int, char**) (.cold) @ 0x396e358 in /home/centos/build/programs/clickhouse
3. /home/centos/ClickHouse/contrib/libcxx/include/vector:461: main @ 0x424b711 in /home/centos/build/programs/clickhouse
4. __libc_start_main @ 0x22555 in /usr/lib64/libc-2.17.so
5. ? @ 0x42d46bf in /home/centos/build/programs/clickhouse`

**Comment by fibersel** [comment]
> How to use it:
> 
> 1. When you will integrate your codecs to `CompressionCodecFactory`, they will become available under symbolic names in clickhouse-compressor.
> 2. You can use clickhouse-compressor to test compression speed and ratio. Then the same for decompression.

How can I measure compression and decompression time most precisely?

**Comment by alexey-milovidov** [comment]
@fibersel `--stat` is for block statistics of compressed file. It does not decompress it and does not show the time.

**Comment by alexey-milovidov** [comment]
> How can I measure compression and decompression time most precisely?

1. Run with `time` command and collect only the `user` value - time in userspace.
2. Run in a loop and collect median or minimum of run times. Minimum can be not robust. If testing on your own machine, set CPU scaling governor to "performance", close web browser, close all other apps that use CPU, pin to one core with numactl.

The difference in various compression algorithms is usually obvious even without these tricks.

**Comment by fibersel** [comment]
Here are first results of testing:
silesia dataset.
Screenshot 2020-12-20 at 20 22 15

**Comment by fibersel** [comment]
Yandex metrics
image

**Comment by alexey-milovidov** [comment]
@fibersel Yandex Metrica should not be used as a single dataset. It contains multiple columns and different columns have very different data distribution. Testing over all columns together will give no insights.

**Comment by alexey-milovidov** [comment]
The results in table are unclear and most likely contain errors:
1. The "level" of lz4 is not specified.
2. If it is the default level (1), decompression speed looks unrealistic and most likely indicates that you have build it in debug mode (e.g. if you are using CLion, it enables debug mode by default).

**Comment by fibersel** [comment]
1. There is not level option for LZ4 into clickhouse-compressor tool.The only option for LZ4 is using HC codec.
2. Is default build type debug?If it is so, I am going to rebuild it...
3. Yes, units are mb/s

**Comment by alexey-milovidov** [comment]
The default build type for ClickHouse is release with debug info (that's what we need).
But if you are using CLion, it has a quirk that sets the build mode to debug by default (that is a common source of confusion).

**Comment by fibersel** [comment]
Hm, no, I build it from command line.

**Comment by fibersel** [comment]
Is it possible that files like URL.dat (600 mb) are too lightweight to measure compression speed?

**Comment by alexey-milovidov** [comment]
600 MB should be enough, it will take hundreds of milliseconds. You can repeat multiple times to get more statistics...

**Comment by alexey-milovidov** [comment]
Actually running multiple times is required for data to fit in page cache (otherwise it can be IO bound).

**Comment by alexey-milovidov** [comment]
Also you can measure only userspace time (`time` will output `user` field).

**Comment by fibersel** [comment]
Sorry for a such long break, but finally I measured first new algorithm. @alexey-milovidov does that representation satisfies you requirements?If it's so, I'll submit code for measuring and benchmarks for other data types later.

Current measurement is for Titles.
|           |   compression (mb/s) |   decompression (mb/s) |   rate (compressed size / initial size) |
|-----------|----------------------|------------------------|-----------------------------------------|
| lizard-10 |            441.7     |                2180.89 |                                0.237939 |
| lizard-13 |             86.479   |                2684.17 |                                0.187487 |
| lizard-16 |             51.7335  |                2791.54 |                                0.176585 |
| lizard-19 |              1.60911 |                2326.28 |                                0.166315 |
| lizard-22 |            193.32    |                1517.14 |                                0.1913   |
| lizard-25 |             29.9265  |                1622.99 |                                0.168324 |
| lizard-28 |              1.2575  |                1702.16 |                                0.148866 |
| lizard-31 |            284.851   |                1586.1  |                                0.198668 |
| lizard-34 |             80.4015  |                1836.54 |                                0.177797 |
| lizard-37 |             51.7719  |                1744.71 |                                0.166143 |
| lizard-40 |            289.579   |                1107.75 |                                0.185636 |
| lizard-43 |             62.0342  |                1163.14 |                                0.162803 |
| lizard-46 |             12.7841  |                1268.88 |                                0.156169 |
| lz4       |            532.737   |                2326.28 |                                0.237471 |
| zstd-11   |             34.4295  |                1292.38 |                                0.126935 |
| zstd-14   |              9.22275 |                1292.38 |                                0.122337 |
| zstd-17   |              5.60145 |                1268.88 |                                0.117369 |
| zstd-20   |              2.54535 |                1268.88 |                                0.116618 |
| zstd-5    |            119.297   |                1073.67 |                                0.141384 |
| zstd-8    |             56.8774  |                1224.36 |                                0.129935 |

**Comment by alexey-milovidov** [comment]
@fibersel Yes, that's what we need, except that measurement should be done for each of the data files.
(we should choose a set of about 5..10 data files that most likely have very different data distribution / data locality)

**Comment by alexey-milovidov** [comment]
Also I think we can first integrate a bunch of interesting libraries... and then we will find something interesting.

File: src/Compression/CompressionCodecDensity.cpp [hunk 1]
```diff
@@ -61,27 +61,31 @@ void registerCodecDensity(CompressionCodecFactory & factory)
...
         {
             DENSITY_ALGORITHM algo = CompressionCodecDensity::DENSITY_DEFAULT_ALGO;
+            //std::cerr << arguments << std::endl;
             if (arguments && !arguments->children.empty())
             {
-                if (arguments->children.size() > 1)
+                if (arguments->children.size() != 1)
                     throw Exception(
-                        "Lizard codec must have 1 parameter, given " + std::to_string(arguments->children.size()),
+                        "Deisnty codec must have 1 parameter, given " + std::to_string(arguments->children.size()),
```
Comment (alexey-milovidov) [inline src/Compression/CompressionCodecDensity.cpp:69]: Typo.

**Comment by fibersel** [comment]
Right now I can't understand why, but LZSSE demonstrates very poor performance.
Authors also claim, that codec is targeted at x64 arhitecture:
https://github.com/ConorStokes/LZSSE/blob/master/lzsse2/lzsse2.h#L67

**Comment by alexey-milovidov** [comment]
There are multiple variants: LZSSE2, LZSSE8 with different tradeofs.
It can be slow on compression but good on decompression.
"x64" is the same as "x86_64" or "amd64" - it is 100% of our servers.

**Comment by fibersel** [comment]
@gonzalezjo Can you please help with codecs' parameters a little?
I'm done with these codecs:
https://github.com/inikep/lizard
https://github.com/ConorStokes/LZSSE
https://github.com/centaurean/density

Now I want to adapt these ones:
https://github.com/IlyaGrebnov/libbsc
https://github.com/fusiyuan2010/CSC
https://github.com/richgel999/lzham_codec

But they have much more complicated interface.Maybe, you can suggest some parameters presets?
Clickhouse interface has only one option - compression level.

**Comment by fibersel** [comment]
Here is an extended benchmark on Titles.Where can I place tests with another data types?
This thread is not the best place, I guess.

|                   |   compression (mb/s) |   decompression (mb/s) |   rate (compressed size / initial size) |
|-------------------|----------------------|------------------------|-----------------------------------------|
| lzsse4-4          |             10.0836  |               3172.21  |                                0.174406 |
| lzsse4-13         |             10.1525  |               2791.54  |                                0.174374 |
| lzsse8-7          |              9.63664 |               2684.17  |                                0.175155 |
| lzsse4-7          |             10.0661  |               2684.17  |                                0.174374 |
| lzsse4-16         |             10.1481  |               2684.17  |                                0.174374 |
| lzsse2-13         |              8.90955 |               2684.17  |                                0.173679 |
| lzsse8-13         |              9.67136 |               2684.17  |                                0.175155 |
| lzsse2-16         |              8.96449 |               2584.76  |                                0.173679 |
| lzsse4-10         |             10.2994  |               2492.45  |                                0.174374 |
| lzsse2-4          |              9.03178 |               2492.45  |                                0.173976 |
| lzsse2-10         |              8.78838 |               2492.45  |                                0.173679 |
| lzsse8-10         |              9.56662 |               2492.45  |                                0.175155 |
| lzsse4-1          |             15.4845  |               2406.5   |                                0.199776 |
| lzsse8-4          |              9.54438 |               2326.28  |                                0.175189 |
| lzsse2-7          |              8.89139 |               2326.28  |                                0.173679 |
| lzsse8-1          |             14.2571  |               2251.24  |                                0.200355 |
| lzsse8-16         |              9.63398 |               2251.24  |                                0.175155 |
| lizard-16         |             48.9058  |               2251.24  |                                0.176585 |
| lizard-13         |             79.5764  |               2180.89  |                                0.187487 |
| lz4-1             |            524.726   |               2114.8   |                                0.237471 |
| lzsse2-1          |             18.2215  |               2114.8   |                                0.224245 |
| lizard-19         |              1.53001 |               2052.6   |                                0.166315 |
| lizard-34         |             77.1997  |               1789.45  |                                0.177797 |
| lizard-10         |            425.54    |               1702.16  |                                0.237939 |
| lizard-28         |              1.17699 |               1661.63  |                                0.148866 |
| lizard-25         |             25.8189  |               1586.1   |                                0.168324 |
| lizard-22         |            179.405   |               1550.86  |                                0.191299 |
| lizard-31         |            262.363   |               1550.86  |                                0.198667 |
| lizard-37         |             45.4352  |               1484.86  |                                0.166143 |
| zstd-14           |              8.30519 |               1246.22  |                                0.122337 |
| zstd-11           |             29.3723  |               1224.36  |                                0.126935 |
| zstd-8            |             54.8221  |               1224.36  |                                0.129935 |
| lizard-46         |             11.7608  |               1224.36  |                                0.156169 |
| zstd-17           |              5.21199 |               1144.07  |                                0.117369 |
| lizard-43         |             55.9203  |               1125.62  |                                0.162803 |
| zstd-20           |              2.38235 |               1057.4   |                                0.116618 |
| zstd-5            |            115.163   |               1057.4   |                                0.141384 |
| lizard-40         |            273.68    |               1011.43  |                                0.185636 |
| density-CHAMELEON |            750.414   |                969.285 |                                0.581477 |
| density-CHEETAH   |            462.176   |                652.229 |                                0.382414 |
| density-LION      |            236.571   |                259.437 |                                0.319485 |

**Comment by alexey-milovidov** [comment]
@fibersel We need to prepare a CSV or JSON file with all the results (all codecs and data types), then we will be able to show interactive diagrams like on https://quixdb.github.io/squash-benchmark/

The first analysis that we need is to select pareto-optimal variants (the variants that no other is strictly better).

**Comment by alexey-milovidov** [comment]
Also we need to make this PR mergeable - so all the codecs you've added will be available under experimental flag.
Then not only you but also I can run some tests :)

**Comment by fibersel** [comment]
Okay, I will make these codecs available under flag next week.

**Comment by alexey-milovidov** [comment]
From your table I see that:
- Lizard is very promising (can decompress faster than ZSTD with similar compression ratio);
- LZSSE is good for faster decompression;
- density looks useless.

**Comment by alexey-milovidov** [comment]
There is still a chance that density is specialized for some datasets - that's why we need to test on multiple columns.

**Comment by alexey-milovidov** [comment]
Also I see that negative compression levels of Lizard are not evaluated.

**Comment by fibersel** [comment]
I guess, negative levels are negative only in the documentation:
https://github.com/inikep/lizard/blob/lizard/lib/lizard_compress.h#L86

**Comment by fibersel** [comment]
@alexey-milovidov how can I get value of experimental flag from inside compression codec?
Should I derive codec form `WithContext` class?

**Comment by alexey-milovidov** [comment]
You can check the flag in factory - where the codecs are created.

**Comment by fibersel** [comment]
Yes, but how can I access global settings?

**Comment by alexey-milovidov** [comment]
You can do it similar to the `allow_suspicious_codecs` setting.

**Pull request merged**